Implement window animations

dccsillag opened the pull request:
This adds window animations, and would solve https://github.com/yshui/picom/issues/217. We use spring-based animations -- they are visually pleasing and handle interactions well, compared to parametric interpolation/animation.

This is still a work in progress -- remaining issues are being tracked in my fork's issue tracker.

I feel it's also worth mentioning that this might have quite a bit of code in common with https://github.com/yshui/picom/pull/766 (I haven't looked into it very carefully). However, with this code the work over at https://github.com/yshui/picom/pull/766 should be easily doable (~10 more lines of extra code), while it probably wouldn't work the other way around.
I've also just stumbled into https://github.com/yshui/picom/issues/771#issuecomment-1028199080 -- this PR should make this very easy (because of the changes to the compose backend operation).

dccsillag committed:
Add start of window animation code

dccsillag committed:
Move animation setting code to before setting w->g to w->pending_g

dccsillag committed:
Add window size animation

dccsillag committed:
Fix animations for windows that have window events very frequent

dccsillag committed:
Fix window animation painting to avoid artifacts

dccsillag committed:
Fix failed assertion when using --experimental-backends

dccsillag committed:
Clean up animation code in `paint_preprocess`

dccsillag committed:
Change animation to use spring-based animation instead of interpolating a parametric curve

dccsillag committed:
Change `window-animations` -> `animations` to conform with the rest of the option/variable names

dccsillag committed:
Merge branch 'yshui:next' into implement-window-animations

dccsillag committed:
Add command line arguments for animation config options

dccsillag committed:
Add animation configs to manpage

dccsillag committed:
Add tips for interactively choosing animation parameters in manpage

alexfrederiksen committed:
Adjust animations to look more natural and fix small issues

Adjusts animations to be with respect to the window center for position
interpolation, which makes things less strange looking when window
scaling takes longer than translating. Also clamps animation size to
window size to resolve blurring artifacts behind the window. Also make
animation fly in from a random direction off screen (:

dccsillag committed:
Add start of proper size animation code for the new experimental backends

Right now, we blend the new image with the old image. What remains is to
resize them so that they fully overlap.

dccsillag committed:
Fix blending for non-opaque for size animations

dccsillag committed:
Change ops->compose backend operation to also receive the bottom-right corner in destination, and use that for animations

Though, actually using it for any backend is still a TODO;
for now, they just do the old compose instead.

dccsillag committed:
Add use of dst_{x,y}2 in gl_compose

dccsillag committed:
Fix size animation blending for opaque windows

This is a bit of a workaround, but necessary for now.

dccsillag committed:
Add optional clamping to window animations

dccsillag committed:
Merge branch 'next' into implement-window-animations

dccsillag committed:
Fix position jump on first window animation

Solves #9.

dccsillag committed:
Fix old_win_image resource management

Previously, we could have discarded it without marking it as discarded.
Solves #10.

dccsillag committed:
Fix memory leak with old_win_image resource management

dccsillag committed:
Fix for when a window gets unmapped in the middle of an animation

Previously, the window stayed static in its middle-of-animation
position. Now, when it's unmapped it gets forced to the end of the
animation.

alexfrederiksen committed:
Add config options and documentation (dccsillag)

dccsillag committed:
Merge remote-tracking branch 'alexfrederiksen/implement-window-animations' into implement-window-animations

dccsillag committed:
Fix open window animations

dccsillag committed:
Merge branch 'next' into implement-window-animations

dccsillag committed:
Remove unnecessary window animation convergence check

dccsillag committed:
Fix sudden animation jumps due to negative values of animation_progress

This happened because there is one single frame in which
w->animation_progress may be negative (because
w->animation_inv_og_distance has not been updated yet). In these cases,
we now do not force the window position to its destination.

dccsillag committed:
Remove remaining convergence checking code

It's unnecessary and may cause sudden animation stops.

dccsillag committed:
Add process_image_flags call to paint_preprocess animation code, after resizing the window

dccsillag committed:
Add checks to see if specific geometry has changed in animation code so that we avoid performing unnecessary costly operations

dccsillag committed:
Fix "none" open-window animation

By removing debug code.

dccsillag committed:
Fix code style

dccsillag committed:
Refactor init_animation code

dccsillag committed:
Add 'zoom' open window animation

dccsillag committed:
Add new compose operation to xrender backend

This effectively adds window content interpolation for window animation
for the xrender backend.

dccsillag committed:
Change resize in xrender backend to use nearest interpolation instead of bilinear

dccsillag committed:
Update animation-for-open-window in asciidoc

dccsillag committed:
Add warning regarding performance and clamping to animation-clamping in asciidoc

See https://github.com/dccsillag/picom/issues/16#issuecomment-907631169

dccsillag committed:
Fix flyin -> fly-in in config.c

pijulius committed:
Fixes artifacts by correctly adding old window area to damage

pijulius committed:
Corrected fix applied before.

pijulius committed:
Fix windows to animate on create in when not destroyed

Seems like sometimes windows aren't animated in case like when you open them in the same spot. You can test this doing the following:
1. Right click in Firefox (anywhere on a website) to open the context menu (will show with animation)
2. Right click again but without moving the mouse (will hide the context menu)
3. Right click again but without moving the mouse (will show the context menu without animation)

With adding those flags when the windows gets unmaped (but I imagine not destroyed) it will animate them correctly even if you open them in the same place.

pijulius committed:
Added new slide animations and possibility to define separate animations for menus and transient windows.

This should add slide animations and also possibility to define different animations depending if the window is a menu or a window that is transient of another one (usually dialog windows that you cannot focus out of like "are you sure to delete xxx file" or about widows and so on).

New animations are:
slide-down
slide-up
slide-left
slide-right

and the option to define different animations for menus and transient windows:
animation-for-menu-window = "slide-down";
animation-for-transient-window = "slide-up";

pijulius committed:
Optimize animations so it doesn't do anything in case animation completed for a window.

The new animations kept on running even if there wasn't anything to do and so processor usage went up badly when animations turned on. This change checks if animation process completed for a window and in that case it simply jumps to the next window without doing all the calculations and other things.

For me this changed picom usage to the same as without animation so 50%+ cpu went down to ~9% just like without animations turned on from config.

Also because of this change animation_running=false is being used correctly as before this it was always true because in the for cycle it always set it to true even if it wasn't anything to do.

pijulius committed:
Reset velocity on animation end and animation_time

This should fix animations from starting at the wrong location as w->animation_velocity_... was never reset and because of the spring animation when a window was zoomed (for e.g. maximized and then restored) the next animation like shading became misplaced meaning it started at a wrong location because of the old velocity value.

Also we reset animation_time to 0 in case there was no animation running at all (just like it's for fading) and so we prevent for e.g. in case clamping turned off to start zooming in windows from thousands of times of your screen and freezing/braking everything.

pijulius committed:
Fixed artifacts in xrender too

This seems to fix the artifacts in xrender too now, we simply add w->reg_ignore_valid = false; to all changes when doing animation and that fixes it. Not sure how this does affect performance but so far works just fine for me.

pijulius committed:
Fix moving transparent windows, add fading to newly created windows.

Added fading to newly created widows so animation doesn't start with a small rectangle fully opaque.
Fixed moving transparent windows by not adding old_image overlay as that way it will start flickering and if you move it just so slightly it will keep flickering.
Fixed opacity value for new_image as it has to be always 1 because w->opacity is already applied later on.

pijulius committed:
Merge animation and fading for loop.

Merged animation loop into the already existing loop used by fading.
Fixed animation Segment fault in case clamping wasn't enabled and the window size got very small (for e.g. in case of shading a window) as width and height became negative number.
Fixed animation_progress to not go back down in case of clamping disabled so that we can rely on it being 1 and this fixes old_image opacity to be hidden and then reappear when maximizing windows and animation goes above real window size.
Fixed animation completed check as 1 = 1L isn't always the same, so we instead check now >= 0.999999999
Wrapped fading code around w->opacity != w->opacity_target for some more optimization.

Overall these changes for me bring down picom from 3-4% cpu to 0.7-1% when idling and hopefully don't brake anything.

pijulius committed:
Revert delta_sec and WINTYPE_TOOLBAR check as was just playing around and left there by mistake.

pijulius committed:
Hopefully fixes empty context menus

pijulius committed:
Fixes black windows for Nvidia users

This should fix the black windows we were having in case of Firefox and other apps. It seems like only happened for Nvidia users as the pixmap was freed when binding new pixmaps but old pixmap wasn't freed and so the new binding didn't work and win_image ended up with a black pixmap.

pijulius committed:
Revert the change of additional ps->animation_time = 0L as not much use for it anyway now that I think I get how things work.

pijulius committed:
Added possibility to define animation on window type basis

This will add possibility to define the animation on window type basis (which deprecated the animation-for-menu-window option) and so now you will be able to define animation like this:

wintypes:
{
  menu = { animation = "slide-down"; }
  popup_menu = { animation = "slide-down"; opacity = 0.97; }
  dropdown_menu = { animation = "slide-down"; opacity = 0.97; }
  notification = { animation = "slide-left"; }
  toolbar = { animation = "none"; }
};

The animation-for-transient-window has been kept as transient window isn't a type but a state so that's a separate thing.

NOTE: if "none" defined for animation for a specified window type that window won't be animated at all, the whole animation feature will be disabled for that kind of window so this way you can force specified windows to not animate at all for e.g. toolbars.

pijulius committed:
Correcty fix black windows

This should now correctly fix black windows and not break animation for old_win_image.
Fully opaque windows moving works. Transparent window moving works. Fully opaque windows animation works. Transparent window animation still flickering a bit but working on a fix.

pijulius committed:
Better implementation of previous fix.

This fix for black windows is even better as it doesn't rely on animation progress and so it works even if animation disabled for those windows.
Tracked down the problem to this:

There's an ev_shape_notify() event fired quickly on new windows for e.g. in case of Firefox main menu and ev_shape_notify() sets the win_set_flags(w, WIN_FLAGS_SIZE_STALE); which brakes the new image captured and because this same event also sets w->reg_ignore_valid = false; too we check for this and this way we ignore the cloning of the image and so new image remains intact.

pijulius committed:
Fixes for resizing transparent windows

pijulius committed:
Revert a quick change left by mistake.

pijulius committed:
Remove opacity from win_image as we use w->opacity now.

dccsillag committed:
Merge branch 'next' into implement-window-animations

dccsillag committed:
Merge branch 'next' into implement-window-animations

dccsillag committed:
Fix rounded corners on old pixmap for window animations

dccsillag committed:
Fix previous commit

dccsillag committed:
Fix code style

yshui commented:
Paging @Arian8j2, do you think you can adapt your work to this? @dccsillag maybe you could give them a hand?

dccsillag commented:
Of course! I can tackle it tomorrow.

yshui commented on src/picom.c:
```diff
@@ -280,6 +280,37 @@ static bool run_fade(session_t *ps, struct managed_win **_w, long steps) {
+ * In milliseconds.
+ */
+static double animation_timeout(session_t *ps) {
+	long double now = get_time_ms();
```
In #766 , @Arian8j2 raised concerns about using wall-clock time for animation. Do you think it would be fine?

dccsillag replied:
I think it should be fine -- it appears to be accurate enough, and, in the case of innacuracies, the math for the animation should be robust in that the window will converge to it's final location and overly weird things shouldn't happen. I haven't noticed any issues at all, nor were there any issues reported which had any relation to the fact that we use wall clock time.

On the other hand, I couldn't find any notes by @Arian8j2 in https://github.com/yshui/picom/pull/766 regarding use of wall clock. Where can I find it?

yshui replied:
Here: https://github.com/yshui/picom/pull/766#discussion_r794391499

Arian8j2 replied:
At first of writing transitions i tried to use delta time for calculating animations like anywhere else, but i just seen a huge difference in some frame times and in that point i just gave up and sticked to calculating for 60 fps and latter i changed it to be based on refresh rate but then @yshui mentioned using steps just like fading step i changed the code to work with steps. Maybe using wall-clock was ok and i was just nitpicking, I should take a second look to it.

yshui commented on src/picom.c:
```diff
@@ -627,15 +658,69 @@ static void handle_root_flags(session_t *ps) {
 }
 
-static struct managed_win *paint_preprocess(session_t *ps, bool *fade_running) {
+static struct managed_win *paint_preprocess(session_t *ps, bool *fade_running, bool *animation_running) {
```
Can fading be merged with this animation code?

dccsillag replied:
Pretty sure it can!

yshui replied:
Yes!

yshui commented on src/picom.c:
```diff
@@ -703,12 +703,19 @@ static struct managed_win *paint_preprocess(session_t *ps, bool *fade_running, b
+				// Synchronize new window position, and add damage to it
 				win_on_win_size_change(ps, w);
+				/* win_update_bounding_shape(ps, w); */
+				win_clear_flags(w, WIN_FLAGS_PIXMAP_STALE);
```
Can you explain what's causing the assertion, and why this fixed it?

dccsillag replied:
Well, in `backend.c` there is the following assert: `assert(!(w->flags & WIN_FLAGS_PIXMAP_STALE));`.

Thing is, in `win_on_win_size_change` (called on line 716 above), the stale flag is set: `win_set_flags(w, WIN_FLAGS_IMAGES_STALE);`. So we unset it, right after. This is fine, since we are about to redraw anyways.

yshui reviewed (changes requested):
Thanks for your work! This is huge :smile: 

I just started reading the code, it might take a while. A few overall comments:

1. Can you squash fixes into appropriate commits? If the fix is tricky or could be a common pitfall, you can add some comments in the code and/or the commit message (for example, ae0f82828e2ea6ccc47a6434871d041434e45193 seems to be interesting).
2. Can you look at the coding style checks and fix them?
3. Can you rebase your branch on top of `next`, instead of merging `next` into it? Unless this would be too much work for you.

dccsillag commented:
> Can you squash fixes into appropriate commits? If the fix is tricky or could be a common pitfall, you can add some comments in the code and/or the commit message (for example, https://github.com/yshui/picom/commit/ae0f82828e2ea6ccc47a6434871d041434e45193 seems to be interesting).

Will do!

> Can you look at the coding style checks and fix them?

Yeah, of course. Would just running `clang-format` fix them?

> Can you rebase your branch on top of next, instead of merging next into it? Unless this would be too much work for you.

Will try!

I don't think I'll be able to work much on this stuff today; I'll be back on it partly tomorrow but mostly over the weekend.

yshui commented:
> I don't think I'll be able to work much on this stuff today; I'll be back on it partly tomorrow but mostly over the weekend.

There is no hurry, please take your time. I won't be able to finish reviewing that quickly anyways.

Arian8j2 commented:
> Paging @Arian8j2, do you think you can adapt your work to this? @dccsillag maybe you could give them a hand?

This is huge animation system and that is so cool to be part of it.

> Can you squash fixes into appropriate commits? If the fix is tricky or could be a common pitfall, you can add some comments in the code and/or the commit message (for example, https://github.com/yshui/picom/commit/ae0f82828e2ea6ccc47a6434871d041434e45193 seems to be interesting).  

That is really needed.

@dccsillag great work after all.

philiparvidsson commented:
Hello! Checking in on this for an update. How are things going - can we somehow help you? I'm looking forward to animated windows in Qtile!

philiparvidsson commented:
I never wanted animations anyway! 😊😁😂😭

dccsillag committed:
Fix code style

dccsillag committed:
Fix window size animations with rounded borders

This is a slightly hacky solution in that it does not properly animate
the borders changing; however, what actually happens in the borders is
barely noticeable (especially with non snail-paced animation speeds), so
it seems to be fine.

dccsillag commented:
I'm terribly sorry for the disappearance. Unfortunately, I've been (and still am) extremely busy with personal matters, so I couldn't really tackle this work much. But I won't be abandoning it; I will continue working, though sporadically.

> How are things going - can we somehow help you?

I'll give here a rough outline of the work that remains here (note that this is not static, I might have missed something considering I've been away for a while; also, the order is a little bit fuzzy, especially towards the end):

- [ ] merge most of https://github.com/dccsillag/picom/pull/22
- [ ] https://github.com/dccsillag/picom/issues/30 and https://github.com/dccsillag/picom/issues/31
- [ ] https://github.com/dccsillag/picom/issues/27
- [ ] rebase with `next` branch
- [ ] clean up commit history
- [ ] merge fading and animation code
- [ ] implement @Arian8j2's work on top of this branch

As for helping -- I think I should be able to finish the first item (merging https://github.com/dccsillag/picom/pull/22) still today. Once that's done, PRs would be very welcome; especially for the second and third items (and possibly the sixth?).

dccsillag committed:
Merge branch 'implement-window-animations' into pijulius-implement-window-animations-tomerge

dccsillag committed:
Merge branch 'next' into implement-window-animations

beyond9thousand commented:
Any updates on this?

yshui commented:
I think all we can do is patiently waiting for @dccsillag

beyond9thousand commented:
> What will happen with the extra animations in pijulius?

@Suyashtnt workspace switch in/out animations.

Suyashtnt commented:
> > What will happen with the extra animations in pijulius?
> 
> @Suyashtnt workspace switch in/out animations.

Yeah I know what they are. I'm asking if they will be added

beyond9thousand commented:
> > > What will happen with the extra animations in pijulius?
> > 
> > 
> > @Suyashtnt workspace switch in/out animations.
> 
> Yeah I know what they are. I'm asking if they will be added

Yes

Alex-vZyl commented:
Imo this PR will finally sort out the picom forking problem.

acama commented:
> Imo this PR will finally sort out the picom forking problem.

Only issue is that we might never see this actually get merged :(

beyond9thousand commented:
Do not lose hope padawan @dccsillag is surely watching over us

dccsillag commented:
> Do not lose hope padawan @dccsillag is surely watching over us

Indeed, I am :smile:
I hope to clock in some work on this next week.

> Only issue is that we might never see this actually get merged :(

I honestly find that quite unlikely (but it *may* take a little while to get merged, unfortunately).

> What will happen with the extra animations in pijulius?

The current plan is to implement them in a second PR -- i.e., this one will contain "only" the actual image transitions and all the backbone of the animation code.
This is because workspace animations, as currently implemented in pijulius' fork, are inconsistent and have a significant memory leak, and solving these problems seemed highly nontrivial. By doing things like this, people will be able to enjoy animations without having to wait even more for all that.

IvanMalison commented:
@dccsillag Is there a good reason for people willing to compile from source not to build from the branch of this pr and use as a daily driver?

dccsillag commented:
> @dccsillag Is there a good reason for people willing to compile from source not to build from the branch of this pr and use as a daily driver?

If any of [the current issues](https://github.com/dccsillag/picom/issues) affect you, then that would be a good reason. Other than that, nothing that comes to mind. In fact, I do this (though mainly for [dogfooding](https://en.wikipedia.org/wiki/Eating_your_own_dog_food)), and I believe someone even made an AUR package.

beyond9thousand commented:
hey, hope you've been doing alright! have you had any luck so far? patiently waiting regardless :D

Suyashtnt commented:
Window animations seem to break severely on awesomeWM, does anybody know the cause of it?

dccsillag commented:
> hey, hope you've been doing alright! have you had any luck so far? patiently waiting regardless :D

I'm afraid I've had no luck. I'm sorry for the inconvenience, but I've really had little to no time to work on this.

> Window animations seem to break severely on awesomeWM, does anybody know the cause of it?

If it's not already in [the issue tracker](https://github.com/dccsillag/picom/issues), then please open one, explaining what you mean by "break severely" (possibly with screenshots/screen recordings).

I recently hapenned to test this on a plain-ish AwesomeWM config and it worked flawlessly -- perhaps you have some configuration causing issues conflicts (but please still report the issue).

Alex-vZyl commented:
@dccsillag  Please don't feel pressured.  This PR does not put food on your table.  Do it in your time.

Borber commented:
I'm looking forward to it! I hope that my phobia of choice will be cured by the official version!

joaopauloalbq commented:
Guys just send messages that have some relevance, this is a pull request not a chat room. 

I also want this feature but this kind of thing just annoys developers...

dccsillag commented:
Just to add to @joaopauloalbq's comment above:
There's no need to worry about me having forgotten about the work here or giving up on it. It's consistently in my personal TODO lists, and if anything is to change about my plans regarding this PR I will clearly say so here.

So let's try to leave this PR's discussion to messages regarding the actual code. (While I don't personally mind comments asking for updates all that much [as long as they are in moderation, as they have generally been], keep in mind that many other people are also subscribed to this PR, and such comments can be very noisy to them.)

soanvig commented:
I've noticed that with `vsync=true` (default I guess?) picom configuration on my setup (4k 60Hz, GTX3080 proprietary driver, experimental backends, glx, kawase blur, rounded corners, shadows) the animation often is laggy (tearing).

However with vsync disabled it's much more smooth.

Monsterovich commented:
@dccsillag Could you please update this PR with "next" branch? I just wanted to take a look into this.

soanvig commented:
FYI: I'm using `implement-window-animations` branch from @dccsillag repo for a month now, and I have no issues.

Monsterovich commented:
> FYI: I'm using `implement-window-animations` branch from @dccsillag repo for a month now, and I have no serious issues. Some slight things could be potentially improved, but it looks very good for a v1 of this feature.

The "next" branch has some important fixes including fixes for memory leaks.

dccsillag commented:
> @dccsillag Could you please update this PR with "next" branch? I just wanted to take a look into this.

I'm waiting on @FT-Labs for that (as per the discussion in https://github.com/yshui/picom/issues/911). BTW, @FT-Labs, any prediction on when we can get a PR on my fork rolling? (Once we have that, I can also start pitching in to help.)

dccsillag commented:
I think you can leave the "flickering" issue (it's more of a pixmap blending issue, really) to me -- I have a couple of ideas on how to at least work around it. So maybe remove the program-specific workaround and once we get a PR going I can tackle it.

Now, the screen center issue is trickier. My guess is that there isn't really going to be some way other than just using Xinerama -- and even if there is, it's probably such a hack that we'd be better off just accepting we have one more dependency now. But I'm not the one to check whether to add a dependency or not, that's up to @yshui.
Nevertheless, here's what I propose regarding this: let's start without these monitor-sensitive animations. Once we have all the rest wrapped up, we can look into adding them back (possibly even as a follow-up PR for yshui/picom).

NicTanghe commented:
is it dead ?

absolutelynothelix commented:
_a wild absolutelynothelix appeared_

@FT-Labs, @dccsillag, xinerama is gone and we're using the x randr extension now (986b3c1cb3b4771eb35de04e7f8b284ce781d208). it's more powerful than you think and we're already getting the dimensions of all monitors, furthermore, there is a function that can tell if a window is entirely on a particular monitor. see `x_update_randr_monitors` and `win_update_monitor` functions in the x.c. you can utilize these mechanism for your needs, it's okay to use the x randr extension.

FT-Labs commented:
@absolutelynothelix This is perfect! Thank you.

https://github.com/FT-Labs/picom/commit/36e6b73cdef2905027df52c58b45fcf7504ee732

FT-Labs commented:
@absolutelynothelix For this to work properly with animations, I needed to delete first statement in the update function.
```
void win_update_monitor(int nmons, region_t *mons, struct managed_win *mw) {
	mw->randr_monitor = -1;
```
Will deleting this assignment break anything? If this exists, animations won't work properly because monitor will get reset on each move.

absolutelynothelix commented:
@FT-Labs, the current implementation of the `win_update_monitor` function assigns a monitor number to a window only if a window is _entirely_ on a particular monitor. if a window is present on multiple monitors, the assigned monitor number will be -1. i haven't looked at your fork yet, but you may want to always assign a monitor number to a window, for example, by calculating intersection areas of a window with each monitor a window is present on, comparing them and assigning a monitor number which has the largest intersection area. in the example below the window 1 will be assigned to the monitor 4 because it has the largest intersection area with it:
```
┌─────────────────────┬─────────────────────┐
│monitor 1            │            monitor 2│
│                     │                     │
│                     │                     │
│                     │                     │
│               ┌─────┼──────────────┐      │
│               │     │              │      │
│               │     │              │      │
├───────────────┼─────┼──────────────┼──────┤
│               │     │              │      │
│               │     │              │      │
│               │     │              │      │
│               │     │      window 1│      │
│               └─────┼──────────────┘      │
│                     │                     │
│monitor 3            │            monitor 4│
└─────────────────────┴─────────────────────┘
```
however, i'm not sure what we should do if intersection areas with two or more monitors are equal.

afaik, the `win_update_monitor` function is used only by the `crop-shadow-to-monitor` option, so you could test if your changes are breaking it. an example how it should look is in the #995.

FT-Labs commented:
@absolutelynothelix I see, that is probably why the issue occurs.

All of the enhanced DE (e.g gnome, kde...) tends to minimize (iconify, unmap event) other windows when switching to another desktop/workspace. However this might not be the case for some window managers. For example, in dwm, there is no unmap/minimize or iconify event. Windows only get hidden to someplace in the display outside of monitor positions. For example they either get hidden on to leftmost monitor - client width, or in y position 0 - client height, without unmapping it. Do you have any idea how we can overcome this issue? Deleting first statement works because it won't get overridden by -1 if client was hidden to any place, but before it has an assigned monitor.

absolutelynothelix commented:
@FT-Labs, try rewriting the `win_update_monitor` function implementing the mechanism described above and not assigning the monitor number -1 to a window in the beginning. it _should_ be fine both for the `crop-shadow-to-monitor` option and for animations, but i can't test it right now. if you don't have multiple monitors to test if it breaks the `crop-shadow-to-monitor` option, you can use the `xrandr` to split a single physical monitor into two or more virtual monitors.

alexhulbert commented:
I'm probably like the only person to extensively use the picom `window-shader-fg` option, but I noticed it doesn't work during and after transitions. The issue persists for about 1-3 seconds after the transition ends (or longer if I don't focus the window).

yshui commented:
@FT-Labs hi, can you open a PR from your branch? I will close the other animation PRs.
